### PR TITLE
Remove `UniformDistribution`, `LogUniformDistribution` and `DiscreteUniformDistribution` code paths

### PR DESCRIPTION
**Pull request by xadrianzetx**
## Motivation

This PR is a continuation to #3166 and #3244. As a part of introducing `FloatDistribution` to the codebase, all code paths taken when instance of old distribution is detected are removed, since logic is now covered in new distribution class.

## Description of the changes

* Remove code paths for `UniformDistribution`, `LogUniformDistribution` and `DiscreteUniformDistribution`
* Fix docstrings

### CODE REVIEW

**Comment by xadrianzetx** [comment]
Note - reason for failing storage tests is explained by https://github.com/optuna/optuna/pull/3166#issuecomment-1009759032. This PR is targeting `feat/unify-distribution` as well.

File: optuna/integration/skopt.py
Comment (himkt) [inline optuna/integration/skopt.py:247]: Sorry after being merged the original PR but could you please copy this comment under `FloatDistribution`?

File: optuna/integration/tensorboard.py
Comment (himkt) [inline optuna/integration/tensorboard.py:59]: [note] We can simply rewrite `supported_distributions` as `[FloatDistribution, IntDistribution, CategoricalDistribution]`, I think. It would be follow-up because cleanup `IntDistribution` is also ongoing.

File: optuna/multi_objective/samplers/_motpe.py
Comment (himkt) [inline optuna/multi_objective/samplers/_motpe.py:40]: In my understanding, `UniformDistribution`, `DiscreteUniformDistribution` and `LogUniformDistribution` are also effective. But I also think it is reasonable to me that only show `FloatDistribution` (not showing deprecated distributions). I'd like other developer's opinion.

https://github.com/optuna/optuna/pull/3275/files#diff-01b5f9e9ffd93b1e908f80f16fae814f065d060e908a779670e905dc30109007R232

File: optuna/samplers/_cmaes.py
Comment (himkt) [inline optuna/samplers/_cmaes.py:63]: :+1:

File: optuna/samplers/_tpe/multi_objective_sampler.py
Comment (himkt) [inline optuna/samplers/_tpe/multi_objective_sampler.py:36]: The same discussion as optuna/multi_objective/samplers/_motpe.py here.

File: optuna/samplers/_tpe/sampler.py
Comment (himkt) [inline optuna/samplers/_tpe/sampler.py:97]: Same as optuna/multi_objective/samplers/_motpe.py. ☝️

File: tests/test_transform.py
Comment (himkt) [inline tests/test_transform.py:95]:
```suggestion
        (0.0, FloatDistribution(0, 3)),
        (3.0, FloatDistribution(0, 3)),
        (1.0, FloatDistribution(1, 10, log=True)),
        (10.0, FloatDistribution(1, 10, log=True)),
        (0.2, FloatDistribution(0, 1, step=0.2)),
        (1.0, FloatDistribution(0, 1, step=0.2)),
```

How about align group by distribution?

File: tests/test_transform.py
Comment (himkt) [inline tests/test_transform.py:170]: I will ask you to resolve conflicts after merging #3181. :pray:

File: tests/samplers_tests/test_samplers.py
Comment (himkt) [inline tests/samplers_tests/test_samplers.py:174]: It is seemingly possible to unify `test_uniform` and `test_log_uniform`. How do you feel to merge them?
If we keep them separated, I prefer to add more test cases or stop using parameterization. 🙋

File: tests/samplers_tests/test_samplers.py
Comment (himkt) [inline tests/samplers_tests/test_samplers.py:149]: It beyonds this PR but I noticed these tests check if samples exist in `[low, upper)`, which doesn't match their documentations. 👀 

https://optuna.readthedocs.io/en/stable/reference/generated/optuna.distributions.UniformDistribution.html#optuna.distributions.UniformDistribution.high

File: tests/samplers_tests/test_samplers.py
Comment (himkt) [inline tests/samplers_tests/test_samplers.py:816]: I'll ask you to resolve conflicts after merging #3181. :bow:

**Review by himkt — COMMENTED** [review]
Thank you @xadrianzetx for the PR! I took a look. :eyes: 📝

File: optuna/multi_objective/samplers/_motpe.py
Comment (HideakiImamura) [reply optuna/multi_objective/samplers/_motpe.py:40]: I think it would be better to just mention FloatDistribution as in the current change; it would help users who have checked the document to reduce the chances of using the old distribution as much as possible.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! Basically, LGTM.

It seems that the following files still have past distributions, mainly in the docstrings. Could you replace them as `FloatDistribution`?
- `optuna/cli.py`
- `optuna/integration/sklearn.py`
- `optuna/study/study.py`
- `optuna/trial/_frozen.py`
- `optuna/tutorial/20_recipes/005_user_defined_sampler.py`

**Comment by HideakiImamura** [comment]
By the way, the CI is failing as follows. This is also happening in #3181. Do you have any idea to fix it?

```
tests/storages_tests/rdb_tests/test_storage.py:256: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
optuna/study/study.py:396: in optimize
    show_progress_bar=show_progress_bar,
optuna/study/_optimize.py:79: in _optimize
    progress_bar=progress_bar,
optuna/study/_optimize.py:163: in _optimize_sequential
    trial = _run_trial(study, func, catch)
optuna/study/_optimize.py:264: in _run_trial
    raise func_err
optuna/study/_optimize.py:213: in _run_trial
    value_or_values = func(trial)
tests/storages_tests/rdb_tests/create_db.py:19: in mo_objective_test_upgrade
    x = trial.suggest_float("x", -5, 5)
optuna/trial/_trial.py:159: in suggest_float
    return self._suggest(name, distribution)
optuna/trial/_trial.py:623: in _suggest
    storage.set_trial_param(trial_id, name, param_value_in_internal_repr, distribution)
optuna/storages/_cached_storage.py:238: in set_trial_param
    study_id, trial_id, param_name, param_value_internal, distribution
optuna/storages/_rdb/storage.py:729: in _check_and_set_param_distribution
    ).check_and_add(session)
optuna/storages/_rdb/models.py:354: in check_and_add
    self._check_compatibility_with_previous_trial_param_distributions(session)
optuna/storages/_rdb/models.py:373: in _check_compatibility_with_previous_trial_param_distributions
    distributions.json_to_distribution(self.distribution_json),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dist_old = UniformDistribution(high=5.0, low=-5.0)
dist_new = FloatDistribution(high=5.0, log=False, low=-5.0, step=None)

    def check_distribution_compatibility(
        dist_old: BaseDistribution, dist_new: BaseDistribution
    ) -> None:
        """A function to check compatibility of two distributions.
    
        Note that this method is not supposed to be called by library users.
    
        Args:
            dist_old: A distribution previously recorded in storage.
            dist_new: A distribution newly added to storage.
    
        Raises:
            ValueError:
                If different distribution kinds are set to ``dist_old`` and ``dist_new``,
                or ``dist_old.choices`` doesn't match ``dist_new.choices``
                for :class:`~optuna.distributions.CategoricalDistribution`.
        """
    
        if dist_old.__class__ != dist_new.__class__:
>           raise ValueError("Cannot set different distribution kind to the same parameter name.")
E           ValueError: Cannot set different distribution kind to the same parameter name.

optuna/distributions.py:598: ValueError
```

**Comment by xadrianzetx** [comment]
Thank you for the first comments @HideakiImamura, @himkt. I'll address them over the weekend or shortly after!

**Comment by himkt** [comment]
Finally, #3181 was merged. 🎉  So sorry for the extra chore but please resolve conflicts.

**Comment by himkt** [comment]
(CI failure will be fixed in https://github.com/optuna/optuna/pull/3113. We still cannot resolve the failure in this PR.)

File: tests/samplers_tests/test_samplers.py
Comment (xadrianzetx) [reply tests/samplers_tests/test_samplers.py:816]: :smiling_face_with_tear:

File: optuna/multi_objective/samplers/_motpe.py
Comment (xadrianzetx) [reply optuna/multi_objective/samplers/_motpe.py:40]: Kept the changes from #3181

File: optuna/samplers/_tpe/multi_objective_sampler.py
Comment (xadrianzetx) [reply optuna/samplers/_tpe/multi_objective_sampler.py]: Done by keeping the changes from https://github.com/optuna/optuna/pull/3181!

File: optuna/samplers/_tpe/sampler.py
Comment (xadrianzetx) [reply optuna/samplers/_tpe/sampler.py]: Done by keeping the changes from https://github.com/optuna/optuna/pull/3181!

File: tests/test_transform.py
Comment (xadrianzetx) [reply tests/test_transform.py]: No problem, done!

File: tests/samplers_tests/test_samplers.py
Comment (xadrianzetx) [reply tests/samplers_tests/test_samplers.py:172]: Great catch! I've gone a bit further and merged cases from `test_uniform`, `test_log_uniform` and `test_discrete_uniform` into one `test_float`. Also, there was a line testing `assert np.all(points < distribution.high)`, which is does not seem correct since `FloatDistribution` can include upper range, so i changed that assertion to weak inequality.

File: optuna/integration/skopt.py
Comment (xadrianzetx) [reply optuna/integration/skopt.py:247]: Done in 5d75afc0630dc3d088151cfb4b6a09eba848316a.

File: tests/test_transform.py
Comment (xadrianzetx) [reply tests/test_transform.py]: Done in 8591eda0f59c0b256023653c8b9d1f1ea2b4a6df.

**Comment by xadrianzetx** [comment]
@himkt, @HideakiImamura, all conflicts are now resolved and your initial comments addressed. Sorry for small delay and PTAL.

**Comment by himkt** [comment]
I'll take a look, thank you for revising @xadrianzetx!

File: optuna/integration/cma.py
Comment (himkt) [inline optuna/integration/cma.py:478]: I know you leave the original implementation. But I think we can merge these branches. 🙋

File: optuna/integration/skopt.py
Comment (himkt) [inline optuna/integration/skopt.py:300]: Now we can simply remove this `if` statement, I think (thanks to the cleanup).

before:

```python
            if isinstance(distribution, distributions.FloatDistribution):
                # Type of value is np.floating, so cast it to Python's built-in float.
                value = float(value)

            if isinstance(distribution, distributions.FloatDistribution):
                if distribution.step is not None:
                    value = value * distribution.step + distribution.low
            elif isinstance(distribution, distributions.IntDistribution):
                if distribution.log:
                    value = int(np.round(value))
                    value = min(max(value, distribution.low), distribution.high)
                else:
                    value = int(value * distribution.step + distribution.low)
```

after:

```python
            if isinstance(distribution, distributions.FloatDistribution):
                # Type of value is np.floating, so cast it to Python's built-in float.
                value = float(value)
                if distribution.step is not None:
                    value = value * distribution.step + distribution.low

            elif isinstance(distribution, distributions.IntDistribution):
                if distribution.log:
                    value = int(np.round(value))
                    value = min(max(value, distribution.low), distribution.high)
                else:
                    value = int(value * distribution.step + distribution.low)
```

File: optuna/study/study.py
Comment (himkt) [inline optuna/study/study.py:954]:
```suggestion
```

Hmm, I just found it is not used in anywhere.

File: tests/test_transform.py
Comment (himkt) [inline tests/test_transform.py:184]: I just noticed but, don't we have to specify log=True in L157 (`x6`) to make equivalent to x15?

File: tests/samplers_tests/test_samplers.py
Comment (himkt) [reply tests/samplers_tests/test_samplers.py:739]: Sorry for that, thank you for the cooperation. :pray:

File: tests/samplers_tests/test_samplers.py
Comment (himkt) [inline tests/samplers_tests/test_samplers.py:171]: I just found `points = points`. 💭 I think we can remove it.
@toshihikoyanase Do you have any specific reason to leave it? The change was introduced in:

https://github.com/optuna/optuna/commit/184e867d1a58082d397600da21a2c6d4f6fa4aaa#diff-8b85ccf8a3534b2aa878bce8ece8ce930af410ab178ebc6d80031badbabb5052L108-R73

**Review by himkt — COMMENTED** [review]
Almost LGTM! I left some minor comments. Please take a look @xadrianzetx!

File: tests/test_transform.py
Comment (xadrianzetx) [reply tests/test_transform.py:184]: You''re right. And same goes for `x5` to be equivalent of `x13`.

File: optuna/integration/cma.py
Comment (xadrianzetx) [reply optuna/integration/cma.py:478]: You mean the ones in `_to_optuna_params`? I'm not sure, they seem to differ quite a bit between `FloatDistribution` and `IntDistribution`.

File: optuna/integration/skopt.py
Comment (xadrianzetx) [reply optuna/integration/skopt.py]: Done in 06d6c8d5ffc729d510e15cb7493a8b14e7c07fce!

File: optuna/study/study.py
Comment (xadrianzetx) [reply optuna/study/study.py]: Good catch! 922752436d2df76bf7e5d5b105ea9640e6e673da.

File: tests/samplers_tests/test_samplers.py
Comment (xadrianzetx) [reply tests/samplers_tests/test_samplers.py:739]: No problem :D

File: optuna/integration/cma.py
Comment (himkt) [reply optuna/integration/cma.py:478]: Ah, sorry for the ambiguous comment. No I don't. I mean I think we can merge two if blocks look like:

### Before

```python
if isinstance(dist, FloatDistribution):
    ...
if isinstance(dist, IntDistribution):
    ...
```

### After

```python
if isinstance(dist, FloatDistribution):
    ...
elif isinstance(dist, IntDistribution):
    ...
```

File: tests/samplers_tests/test_samplers.py
Comment (himkt) [reply tests/samplers_tests/test_samplers.py:171]: I asked @toshihikoyanase and confirmed it leaves unintentionally. Please remove it as well. :pray:

Comment (xadrianzetx) [reply tests/samplers_tests/test_samplers.py]: Done in 2b8034e9d4710ea9fea3661371d295fce963d9ea!

File: optuna/integration/cma.py
Comment (xadrianzetx) [reply optuna/integration/cma.py]: Ah, yes, makes sense, thank you for explaining and done in 6abd4fcefc408f020a5252544cc3efae0b3f49ec.

**Review by himkt — APPROVED** [review]
LGTM!!

I finally got the tests green on #3113! (after rebasing)

```
> python -m pytest tests/storages_tests/rdb_tests/test_storage.py                                                                                                    2022-02-09 22:23:31
================================================================================== test session starts ==================================================================================
platform darwin -- Python 3.9.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/himkt/work/github.com/himkt/optuna, configfile: pyproject.toml
collected 24 items

tests/storages_tests/rdb_tests/test_storage.py ........................                                                                                                           [100%]
...
============================================================================ 24 passed, 44 warnings in 6.89s ============================================================================
(venv) [x86_64] [~/work/github.com/himkt/optuna] @ migrate/distribution ●
```

**Comment by xadrianzetx** [comment]
Nice! We are getting close to merge this back to master @himkt! I can see there are some outstanding tasks in https://github.com/optuna/optuna/pull/3166#pullrequestreview-857824037. Should i tackle them before we merge back to master?

**Comment by himkt** [comment]
I prefer to apply comments before merging master, but it would be easier to work on after this PR. 🙋 
The changes looks solid but it could require additional changes by another review.

**Comment by xadrianzetx** [comment]
Yeah, so we merge this one, then #3113 to `feat/unify-distribution`, then minor stuff from the comments and after that we can merge all this to master, right?

**Review by HideakiImamura — APPROVED** [review]
LGTM! Thanks for your contribution!

**Pull request merged**
